src/ctc_core.c

```diff
@@ -0,0 +1,462 @@
+#include <stdlib.h>
+#include <regex.h>
+#include "ctc_core.h"
+
+pthread_once_t ctc_api_once_init = PTHREAD_ONCE_INIT;
+
+CTC_HANDLE ctc_pool[MAX_CTC_HANDLE_COUNT];
+
+void ctc_api_init (void)
+{
+    int i, j;
+
+    for (i = 0; i < MAX_CTC_HANDLE_COUNT; i ++)
+    {
+        ctc_pool[i].ID = i;
+        ctc_pool[i].control_session.session_gid = -1;
+
+        for (j = 0; j < MAX_JOB_HANDLE_COUNT; j ++)
+        {
+            ctc_pool[i].job_pool[j].ID = j;
+            ctc_pool[i].job_pool[j].job_session.job_desc = -1;
+        }
+    }
+}
+
+int alloc_ctc_handle (CTC_HANDLE **ctc_handle_p)
+{
+    int i;
+
+    *ctc_handle_p = NULL;
+
+    for (i = 0; i < MAX_CTC_HANDLE_COUNT; i ++)
+    {
+        if (ctc_pool[i].control_session.session_gid == -1)
+        {
+            *ctc_handle_p = &ctc_pool[i];
+            
+            break;
+        }
+    }
+
+    if (IS_NULL (*ctc_handle_p))
+    {
+        goto error;
+    }
+
+    return CTC_SUCCESS;
+
+error:
+
+    return CTC_FAILURE;
+}
+
+int free_ctc_handle (CTC_HANDLE *ctc_handle)
+{
+    // 소켓 정리, 할당된 job 정리
+
```

Comment:
Null value testing for argument 'ctc_handle' is recommended before proceed. 
Applications may call library functions improper order.
It is better to check other library functions as well.

Reply:
The free_ctc_handle function is never called by Apps directly. And I think that the arguments from App already are validated in find_ctc_handle function.